### PR TITLE
[ENG-613] Indexer rule whitelist/blacklist per rule

### PR DESCRIPTION
**Pull request by fogodev**
This PR introduces a semantic change of how Indexer Rules are represented, before we had that each Indexer Rule could have just one of the available kinds. Now each Indexer Rule can have many rules of each kind, according changes were made in the walker to enable this feature.

### CODE REVIEW

File: core/src/location/indexer/walk.rs [hunk 1]
```diff
@@ -780,11 +744,17 @@ mod tests {
 			420,
 		)
 		.await
-		.unwrap()
-		.walked
-		.collect::<HashSet<_>>();
+		.unwrap();
+
+		if !walk_result.errors.is_empty() {
+			panic!("errors: {:#?}", walk_result.errors);
```
Comment (brxken128) [inline core/src/location/indexer/walk.rs:750]: is `panic!`ing the best thing we can do here? seems unfitting given all of the new error handling stuff

**Review by brxken128 — COMMENTED** [review]
looks pretty good to me, just a little curious about the `panic!`s (they may be fine though tbh) 🙏🏻

**Comment by HeavenVolkoff** [comment]
Migration seems to be broken:
image

**Comment by HeavenVolkoff** [comment]
Tested here and everything seems to be working fine, except for the migration issue

File: core/src/location/indexer/walk.rs (hunk 1 above)
Comment (fogodev) [reply core/src/location/indexer/walk.rs:750]: These are on unit tests, so I think it's alright

**Comment by fogodev** [comment]
> Migration seems to be broken: image

Looks like this one is caused by the "migrator" on dev builds, and not by the migrations files. As I actually introduced a breaking change on how indexer rules are stored on database, now storing bytes representing a vector of enums for each rule kind, it isn't backward compatible anyway, so you have to remove your database. Otherwise we would have to start handling data migrations for incompatible data

**Comment by HeavenVolkoff** [comment]
> Looks like this one is caused by the "migrator" on dev builds, and not by the migrations files. As I actually introduced a breaking change on how indexer rules are stored on database, now storing bytes representing a vector of enums for each rule kind, it isn't backward compatible anyway, so you have to remove your database. Otherwise we would have to start handling data migrations for incompatible data

Hun... this is interesting. Prod would also require the user to delete the db, right? Maybe we should handle data migration then. In this specific case, maybe just dropping the old table would be enough. @Brendonovich what you think?

**Comment by fogodev** [comment]
> > Looks like this one is caused by the "migrator" on dev builds, and not by the migrations files. As I actually introduced a breaking change on how indexer rules are stored on database, now storing bytes representing a vector of enums for each rule kind, it isn't backward compatible anyway, so you have to remove your database. Otherwise we would have to start handling data migrations for incompatible data
> 
> Hun... this is interesting. Prod would also require the user to delete the db, right? Maybe we should handle data migration then. In this specific case, maybe just dropping the old table would be enough. @Brendonovich what you think?

Nope, as on prod we actually use the migrations, and in the migrations the old indexer rules table is dropped.

**Comment by HeavenVolkoff** [comment]
> Nope, as on prod we actually use the migrations, and in the migrations the old indexer rules table is just dropped, without trying to migrate the old incompatible data. And we also repopulate the indexer rules table with basic rules in the seeder, using the new data format.

But that would mean the user will lose his custom indexer rules, no? What would happen if there are locations using them?

File: core/src/location/indexer/walk.rs (hunk 1 above)
Comment (brxken128) [reply core/src/location/indexer/walk.rs:750]: that is a great point haha

**Comment by fogodev** [comment]
> > Nope, as on prod we actually use the migrations, and in the migrations the old indexer rules table is just dropped, without trying to migrate the old incompatible data. And we also repopulate the indexer rules table with basic rules in the seeder, using the new data format.
> 
> Actually, dropping isn't a good solution, as it would mean the user will lose his custom indexer rules. What would happen if there are locations using them?

Yeah, if we already had a stable release to maintain, I would totally agree that custom indexer rules are important, and we should have some kind of data migration to keep them. But it isn't the case yet.

**Comment by Brendonovich** [comment]
Gonna chime in to say that right now migrations aren't important and we can mangle them as much as we want, they'll be squashed into a single file once we go stable anyway

**Review by HeavenVolkoff — APPROVED** [review]
Tested again after the merge with main. Everything seems to be working fine. Not auto-merging yet just to wait the mobile test to pass, and also if someone else wants to review it...

**Pull request merged**